packages/xod-client/src/editor/components/inspectorWidgets/index.js

```diff
@@ -56,10 +56,35 @@ const WIDGET_MAPPING = {
     props: {
       type: PIN_TYPE.NUMBER,
       keyDownHandlers: R.merge(widgetNumberKeysDownHandlers, submitOnEnter),
+      normalizeValue: R.compose(
+        // If number is not valid C++ number with scientific notation
+        // fallback to '0'.
+        R.unless(
+          R.test(/^(-|\+)?\d+(\.\d+)?((e-|e\+|e|\d+)\d+)?$|^NaN$|^(-|\+)Inf$/),
+          R.always('0')
+        ),
+        // Let user type `nan`, `+inf`, `-inf` values with wrong case
+        // and type `+Inf` without plus symbol — correct it automatically
+        R.cond([
+          [R.test(/^nan$/i), R.always('NaN')],
+          [R.test(/^\+?inf$/i), R.always('+Inf')],
+          [R.test(/^-inf$/i), R.always('-Inf')],
+          [R.T, R.identity],
+        ]),
+        ensureLiteral(PIN_TYPE.NUMBER)
```

Comment:
A simple lowercasing before the `cond` will make things better: no need for regex testing, a `42E+3` will be converted to `42e+3` automatically.